trajectory_executor/com_serial.py

```diff
@@ -0,0 +1,222 @@
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_KNEE']
+        elif read_state == serial_read_states['READ_R_KNEE']:
+          data_packet['r_knee'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_L_SHOULDER']
+        elif read_state == serial_read_states['READ_L_SHOULDER']:
+          data_packet['l_shoulder'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_L_ELBOW']
+        elif read_state == serial_read_states['READ_L_ELBOW']:
+          data_packet['l_elbow'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_SHOULDER']
+        elif read_state == serial_read_states['READ_R_SHOULDER']:
+          data_packet['r_shoulder'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_R_ELBOW']
+        elif read_state == serial_read_states['READ_R_ELBOW']:
+          data_packet['r_elbow'] += decoded_data
+          calculated_checksum += decoded_data
+          data_byte_counter -= 1
+          if data_byte_counter == 0:
+            data_byte_counter = self.data_byte_length
+            read_state = serial_read_states['READ_AT_GOAL']
+        elif read_state == serial_read_states['READ_AT_GOAL']:
+          data_packet['at_goal'] = decoded_data
+          calculated_checksum += decoded_data
+          read_state = serial_read_states['READ_CHECKSUM']
+        elif read_state == serial_read_states['READ_CHECKSUM']:
+          data_packet['checksum'] = decoded_data
+          if data_packet['checksum'] == 255 - (calculated_checksum % 256):
+            # good packet, relay to tracking node
+            response.roll = data_packet['roll']
+            response.pitch = data_packet['pitch']
+            response.yaw = data_packet['yaw']
+            response.l_hip = data_packet['l_hip']
+            response.l_knee = data_packet['l_hip']
+            response.r_hip = data_packet['r_hip']
+            response.r_knee = data_packet['r_hip']
+            response.l_shoulder = data_packet['l_shoulder']
+            response.l_elbow = data_packet['l_elbow']
+            response.r_shoulder = data_packet['r_shoulder']
+            response.r_elbow = data_packet['r_elbow']
+            response.at_goal = data_packet['at_goal']
```

Comment:
How does ROS handle these automatically generated `Request` objects? Basically, can you instantiate it and set the values at construction time--ie. `response = SensorDataRequest(roll=<roll>, pitch=<pitch>, ...)`?

If so, then you can use keyword expansion and do all of this in one swoop with 
```python
response = SensorDataRequest(**data_packet)
```

Reply:
can also do
```python
for attr, value in data_packet.items():
   setattr(response, attr, value)
```
with the assumption that the keys in `data_packet` correspond to the attributes in `response`

Reply:
I'm not sure. In the tutorials, they've always created the object first then populated fields. I can post a question on the ROS answers forum and experiment with it a little.

Reply:
iirc the second example should work cause I believe that they are just Python objects. idk feel free to experiment with it. I'm not sure how much the data packet will change down the line but this would make it fully dynamic.

Reply:
I think we can test this out with a simple service example from ROS. I like the idea of it being dynamic

Reply:
Yea there are ways to do various testing on it. For now let's leave it as is and make a ticket to revisit this